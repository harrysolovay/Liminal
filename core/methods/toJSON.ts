import { isType } from "../isType.ts"
import type { JSONType } from "../JSONSchema.ts"
import type { PartialType, Type } from "../Type.ts"
import { TypeVisitor } from "../TypeVisitor.ts"
import { DescriptionContext } from "./description.ts"

export function toJSON(this: Type<unknown, any>): JSONType {
  const ctx = new VisitorContext(
    new Map(),
    {},
    new DescriptionContext(new Map(), {}),
  )
  const root = visit(ctx, this)
  const { "0": _root, ...$defs } = ctx.defs
  return { ...root, $defs } as never
}

class VisitorContext {
  constructor(
    readonly ids: Map<PartialType, string>,
    readonly defs: Record<string, undefined | JSONType>,
    readonly descriptionCtx: DescriptionContext,
  ) {}

  id(type: PartialType): string {
    let id = this.ids.get(type)
    if (id === undefined) {
      id = this.ids.size.toString()
      this.ids.set(type, id)
    }
    return id
  }
}

const visit = TypeVisitor<VisitorContext, JSONType>({
  hook(next, { descriptionCtx: { args, pins }, ids, defs }, type) {
    let jsonType: JSONType
    args = { ...args }
    const descriptionCtx = new DescriptionContext(pins, args)
    const description = descriptionCtx.format(type as never)
    const ctx = new VisitorContext(ids, defs, descriptionCtx)
    if (isType(type, "array", "object", "union")) {
      const id = ctx.id(type)
      if (id in defs) {
        return defs[id] ?? {
          $ref: id === "0" ? "#" : `#/$defs/${id}`,
        }
      } else {
        defs[id] = undefined
        jsonType = next(ctx, type)
        defs[id] = jsonType
      }
    } else {
      jsonType = next(ctx, type)
    }
    return { description, ...jsonType }
  },
  null() {
    return { type: "null" }
  },
  boolean() {
    return { type: "boolean" }
  },
  integer() {
    return { type: "integer" }
  },
  number() {
    return { type: "number" }
  },
  string() {
    return { type: "string" }
  },
  const(ctx, _1, valueType, value): JSONType {
    return {
      ...visit(ctx, valueType),
      const: value,
    }
  },
  array(ctx, _1, element): JSONType {
    return {
      type: "array",
      items: visit(ctx, element),
    }
  },
  object(ctx, _1, fields): JSONType {
    const required = Object.keys(fields)
    return {
      type: "object",
      properties: Object.fromEntries(required.map((k) => [k, visit(ctx, fields[k]!)])),
      required,
      additionalProperties: false,
    }
  },
  enum(_0, _1, ...values) {
    return {
      type: "string",
      enum: values,
    }
  },
  union(ctx, _1, ...members): JSONType {
    return {
      anyOf: members.map((member) => visit(ctx, member)),
    }
  },
  ref(ctx, _1, get): JSONType {
    return visit(ctx, get())
  },
  transform(ctx, _1, from): JSONType {
    return visit(ctx, from)
  },
})
