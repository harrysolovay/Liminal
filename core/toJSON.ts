import { DescriptionContext } from "./DescriptionContext.ts"
import * as I from "./intrinsics/mod.ts"
import type { JSONType } from "./JSONSchema.ts"
import { type AnyType, Type } from "./Type.ts"
import { TypeVisitor } from "./TypeVisitor.ts"

export function toJSON(type: Type<unknown, any>): JSONType {
  const ctx = new VisitorContext(
    new Map(),
    {},
    new DescriptionContext(new Map(), {}),
  )
  const root = visit(ctx, type)
  const { "0": _root, ...$defs } = ctx.defs
  return { ...root, $defs } as never
}

class VisitorContext {
  constructor(
    readonly ids: Map<AnyType, string>,
    readonly defs: Record<string, undefined | JSONType>,
    readonly descriptionCtx: DescriptionContext,
  ) {}

  id(type: AnyType): string {
    let id = this.ids.get(type)
    if (id === undefined) {
      id = this.ids.size.toString()
      this.ids.set(type, id)
    }
    return id
  }
}

const visit = TypeVisitor<VisitorContext, JSONType>({
  hook(next, { descriptionCtx: { args, pins }, ids, defs }, type) {
    let jsonType: JSONType
    args = { ...args }
    const descriptionCtx = new DescriptionContext(pins, args)
    const description = descriptionCtx.format(type as never)
    const ctx = new VisitorContext(ids, defs, descriptionCtx)
    if (Type.extends(type, I.array, I.object, I.union)) {
      const id = ctx.id(type)
      if (id in defs) {
        jsonType = defs[id] ?? {
          $ref: id === "0" ? "#" : `#/$defs/${id}`,
        }
      } else {
        defs[id] = undefined
        jsonType = next(ctx, type)
        defs[id] = jsonType
      }
    } else {
      jsonType = next(ctx, type)
    }
    return { description, ...jsonType }
  },
  null() {
    return { type: "null" }
  },
  boolean() {
    return { type: "boolean" }
  },
  integer() {
    return { type: "integer" }
  },
  number() {
    return { type: "number" }
  },
  string() {
    return { type: "string" }
  },
  const(ctx, _1, valueType, value): JSONType {
    return {
      ...visit(ctx, valueType),
      const: value,
    }
  },
  array(ctx, _1, element): JSONType {
    return {
      type: "array",
      items: visit(ctx, element),
    }
  },
  object(ctx, _1, fields): JSONType {
    const required = Object.keys(fields)
    return {
      type: "object",
      properties: Object.fromEntries(required.map((k) => [k, visit(ctx, fields[k]!)])),
      required,
      additionalProperties: false,
    }
  },
  enum(_0, _1, ...values) {
    return {
      type: "string",
      enum: values,
    }
  },
  union(ctx, _1, ...members): JSONType {
    return {
      anyOf: members.map((member) => visit(ctx, member)),
    }
  },
  ref(ctx, _1, get): JSONType {
    return visit(ctx, get())
  },
  transform(ctx, _1, from): JSONType {
    return visit(ctx, from)
  },
})
