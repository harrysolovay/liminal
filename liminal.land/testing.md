# Liminal Testing <Badge type="warning" text="beta" />
