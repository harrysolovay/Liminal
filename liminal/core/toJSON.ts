import { DescriptionContext } from "./description.ts"
import type { JSONType, JSONTypeKind, JSONTypes } from "./JSONSchema.ts"
import type { AnyType, Type } from "./Type.ts"
import { TypeVisitor } from "./TypeVisitor.ts"

export function toJSON<K extends JSONTypeKind>(this: Type<K, unknown, any>): JSONTypes[K] {
  const ctx = new VisitContext(
    new Map(),
    {},
    new DescriptionContext(new Map(), {}),
  )
  const root = visit(ctx, this)
  const { "0": _root, ...$defs } = ctx.defs
  return { ...root, $defs } as never
}

class VisitContext {
  constructor(
    readonly ids: Map<AnyType, string>,
    readonly defs: Record<string, undefined | JSONType>,
    readonly descriptionCtx: DescriptionContext,
  ) {}
}

const visit = TypeVisitor<VisitContext, JSONType>({
  hook(next, { descriptionCtx: { args, pins }, ids, defs }, type) {
    let jsonType: JSONType
    args = { ...args }
    const descriptionCtx = new DescriptionContext(pins, args)
    const description = (type as Type<JSONTypeKind, unknown, never>).description(descriptionCtx)
    const ctx = new VisitContext(ids, defs, descriptionCtx)
    switch (type.K) {
      case "array":
      case "object":
      case "union": {
        let id = ids.get(type)
        if (id === undefined) {
          id = ids.size.toString()
          ids.set(type, id)
        }
        if (id in defs) {
          jsonType = defs[id] ?? {
            $ref: `#/$defs/${id}`,
          }
        } else {
          defs[id] = undefined
          defs[id] = next(ctx, type)
        }
        jsonType = next(ctx, type)
        break
      }
      default: {
        jsonType = next(ctx, type)
        break
      }
    }
    return {
      description,
      ...jsonType,
    }
  },
  null() {
    return { type: "null" }
  },
  boolean() {
    return { type: "boolean" }
  },
  integer() {
    return { type: "integer" }
  },
  number() {
    return { type: "number" }
  },
  string() {
    return { type: "string" }
  },
  const(ctx, _1, valueType, value): JSONType {
    return {
      ...visit(ctx, valueType),
      const: value,
    }
  },
  array(ctx, _1, element): JSONType {
    return {
      type: "array",
      items: visit(ctx, element),
    }
  },
  object(ctx, _1, fields): JSONType {
    const required = Object.keys(fields)
    return {
      type: "object",
      properties: Object.fromEntries(required.map((k) => [k, visit(ctx, fields[k]!)])),
      required,
      additionalProperties: false,
    }
  },
  enum(_0, _1, ...values) {
    return {
      type: "string",
      enum: values,
    }
  },
  union(ctx, _1, ...members): JSONType {
    return {
      anyOf: members.map((member) => visit(ctx, member)),
    }
  },
  ref(ctx, _1, get): JSONType {
    return {
      $ref: `#/$defs/${ctx.ids.get(get())!}`,
    }
  },
  transform(ctx, _1, from): JSONType {
    return visit(ctx, from)
  },
})
