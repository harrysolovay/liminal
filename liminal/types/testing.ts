import * as T from "./index.ts"

const g = T.union(
  T.object({
    a: T.number`Testing`,
    b: T.string,
    c: T.const(T.string, "hello"),
    d: {
      another: T.string,
      nested: T.object({
        e: T.number,
      }),
    },
    x: ["A", "B"],
  })`
  Testing
`,
  T.string,
)

console.log(JSON.stringify(g.toJSON(), null, 2))
