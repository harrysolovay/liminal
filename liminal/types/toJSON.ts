import type { JSONValue } from "../util/JSONValue.ts"
import { _object, type _ObjectFields, type JSONObjectType } from "./_object.ts"
import type { JSONArrayType } from "./array.ts"
import type { JSONBooleanType } from "./boolean.ts"
import { type JSONConstType } from "./const.ts"
import type { JSONEnumType } from "./enum.ts"
import type { JSONIntegerType } from "./integer.ts"
import type { JSONRootType } from "./JSONRootType.ts"
import { JSONSchema$schema } from "./JSONRootType.ts"
import type { JSONRootableType, JSONType } from "./JSONType"
import type { JSONNullType } from "./null.ts"
import type { JSONNumberType } from "./number.ts"
import type { JSONRefType } from "./ref.ts"
import { type JSONStringType } from "./string.ts"
import { type Type } from "./Type.ts"
import { TypeContext } from "./TypeContext.ts"
import { isParentType, TypeVisitor } from "./TypeVisitor.ts"
import type { JSONUnionType } from "./union.ts"

export function toJSON<T extends JSONValue, J extends JSONRootableType>(this: Type<T, J>): JSONRootType<J> {
  const { $defs: { 0: root, ...$defs } } = new ToJSONContext(this)
  return {
    $schema: JSONSchema$schema,
    ...root,
    $defs,
  } as never
}

class ToJSONContext extends TypeContext {
  readonly $defs: Record<string, JSONType | undefined> = {}
  constructor(root: Type) {
    super(root)
    visit(this, root)
  }
}

const visit = TypeVisitor<ToJSONContext, JSONType>({
  hook(next, state, type) {
    if (isParentType(type)) {
      const id = state.id(type)
      if (!(id in state.$defs)) {
        state.$defs[id] = undefined
        state.$defs[id] = next(state, type)
      }
      return type === state.root ? { $ref: "$" } : { $ref: `#/$defs/${id}` }
    }
    return next(state, type)
  },
  null(): JSONNullType {
    return { type: "null" }
  },
  boolean(): JSONBooleanType {
    return { type: "boolean" }
  },
  integer(): JSONIntegerType {
    return { type: "integer" }
  },
  number(): JSONNumberType {
    return { type: "number" }
  },
  string(): JSONStringType {
    return { type: "string" }
  },
  const(state, _type, valueType, value): JSONConstType {
    const initial = visit(state, valueType)
    return Object.assign(initial as Exclude<typeof initial, JSONRefType>, {
      const: value,
    })
  },
  array(state, _type, element): JSONArrayType {
    return {
      type: "array",
      items: visit(state, element),
    }
  },
  enum(_state, _type, ...values): JSONEnumType {
    return {
      type: "string",
      enum: values,
    }
  },
  union(state, _type, ...members): JSONUnionType {
    return {
      anyOf: members.map((member) => visit(state, member)),
    }
  },
  _object(state, _type, fields): JSONObjectType {
    return {
      type: "object",
      properties: Object.fromEntries(
        Array.isArray(fields)
          ? fields.map((value, i) => [i, visit(state, value)])
          : Object.entries(fields).map(([key, value]) => [key, visit(state, value)]),
      ),
      required: Object.keys(fields),
      additionalProperties: false,
    }
  },
  ref(state, _type, getType): JSONRefType {
    return visit(state, getType()) as JSONRefType
  },
})
