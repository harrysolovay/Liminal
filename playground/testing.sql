-- migrate:up
CREATE EXTENSION if NOT EXISTS "uuid-ossp";

CREATE EXTENSION if NOT EXISTS "hstore";

CREATE EXTENSION if NOT EXISTS "vector";

CREATE EXTENSION if NOT EXISTS "lo";

CREATE TABLE IF NOT EXISTS todo (
  id serial PRIMARY KEY,
  task TEXT,
  done BOOLEAN DEFAULT FALSE
);

INSERT INTO
  todo (task, done)
VALUES
  ('Install PGlite from NPM', TRUE);

INSERT INTO
  todo (task, done)
VALUES
  ('Load PGlite', TRUE);

INSERT INTO
  todo (task, done)
VALUES
  ('Create a table', TRUE);

INSERT INTO
  todo (task, done)
VALUES
  ('Insert some data', TRUE);

INSERT INTO
  todo (task)
VALUES
  ('Update a task');
