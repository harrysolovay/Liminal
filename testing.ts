import { ensureDir } from "@std/fs"
import * as path from "@std/path"
import { tap } from "./util/tap.ts"

export const dbg = tap(async (value) => {
  console.debug(value)
  await ensureDir(TMP_DIR)
  const dest = path.join(TMP_DIR, `${Date.now()}.json`)
  await Deno.writeTextFile(dest, JSON.stringify(value, null, 2))
  console.log(`Written to ${dest}.`)
})

const TMP_DIR = ".tmp"
