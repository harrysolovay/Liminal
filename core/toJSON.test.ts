import { assertSnapshot } from "@std/testing/snapshot"
import { testIntrinsics } from "testing"
import { L } from "./mod.ts"
import { toJSON } from "./toJSON.ts"
import type { Type } from "./Type.ts"

testIntrinsics("toJSON", assertTypeSnapshot, {
  null: [
    [L.null],
  ],
  const: [
    [L.const(L.string, "Hello Liminal!")],
  ],
  boolean: [
    [L.boolean],
  ],
  number: [
    [L.number],
  ],
  integer: [
    [L.integer],
  ],
  string: [
    [L.string],
  ],
  array: [
    [L.array(L.string)],
  ],
  object: [
    [
      L.object({
        a: L.boolean,
        b: L.number,
        c: L.string,
      }),
    ],
  ],
  enum: [
    [L.enum("A", "B", "C")],
  ],
  union: [
    [L.union(L.null, L.string)],
  ],
  ref: [
    [L.ref(() => L.string)],
  ],
  transform: [
    [L.transform(L.string, (value) => ({ value }))],
  ],
})

async function assertTypeSnapshot(t: Deno.TestContext, value: Type<unknown, never>): Promise<void> {
  await assertSnapshot(t, toJSON(value))
  const withContext = value`One.`
  await assertSnapshot(t, toJSON(withContext))
  await assertSnapshot(t, toJSON(withContext`Two.`))
}
