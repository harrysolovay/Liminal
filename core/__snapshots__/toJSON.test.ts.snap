export const snapshot = {};

snapshot[`toJSON > null 1`] = `
{
  "\$defs": {},
  description: undefined,
  type: "null",
}
`;

snapshot[`toJSON > null 2`] = `
{
  "\$defs": {},
  description: "One.",
  type: "null",
}
`;

snapshot[`toJSON > null 3`] = `
{
  "\$defs": {},
  description: "One. Two.",
  type: "null",
}
`;

snapshot[`toJSON > const 1`] = `
{
  "\$defs": {},
  const: "Hello Liminal!",
  description: undefined,
  type: "string",
}
`;

snapshot[`toJSON > const 2`] = `
{
  "\$defs": {},
  const: "Hello Liminal!",
  description: undefined,
  type: "string",
}
`;

snapshot[`toJSON > const 3`] = `
{
  "\$defs": {},
  const: "Hello Liminal!",
  description: undefined,
  type: "string",
}
`;

snapshot[`toJSON > boolean 1`] = `
{
  "\$defs": {},
  description: undefined,
  type: "boolean",
}
`;

snapshot[`toJSON > boolean 2`] = `
{
  "\$defs": {},
  description: "One.",
  type: "boolean",
}
`;

snapshot[`toJSON > boolean 3`] = `
{
  "\$defs": {},
  description: "One. Two.",
  type: "boolean",
}
`;

snapshot[`toJSON > number 1`] = `
{
  "\$defs": {},
  description: undefined,
  type: "number",
}
`;

snapshot[`toJSON > number 2`] = `
{
  "\$defs": {},
  description: "One.",
  type: "number",
}
`;

snapshot[`toJSON > number 3`] = `
{
  "\$defs": {},
  description: "One. Two.",
  type: "number",
}
`;

snapshot[`toJSON > integer 1`] = `
{
  "\$defs": {},
  description: undefined,
  type: "integer",
}
`;

snapshot[`toJSON > integer 2`] = `
{
  "\$defs": {},
  description: "One.",
  type: "integer",
}
`;

snapshot[`toJSON > integer 3`] = `
{
  "\$defs": {},
  description: "One. Two.",
  type: "integer",
}
`;

snapshot[`toJSON > string 1`] = `
{
  "\$defs": {},
  description: undefined,
  type: "string",
}
`;

snapshot[`toJSON > string 2`] = `
{
  "\$defs": {},
  description: "One.",
  type: "string",
}
`;

snapshot[`toJSON > string 3`] = `
{
  "\$defs": {},
  description: "One. Two.",
  type: "string",
}
`;

snapshot[`toJSON > array 1`] = `
{
  "\$defs": {},
  description: undefined,
  items: {
    description: undefined,
    type: "string",
  },
  type: "array",
}
`;

snapshot[`toJSON > array 2`] = `
{
  "\$defs": {},
  description: "One.",
  items: {
    description: undefined,
    type: "string",
  },
  type: "array",
}
`;

snapshot[`toJSON > array 3`] = `
{
  "\$defs": {},
  description: "One. Two.",
  items: {
    description: undefined,
    type: "string",
  },
  type: "array",
}
`;

snapshot[`toJSON > object 1`] = `
{
  "\$defs": {},
  additionalProperties: false,
  description: undefined,
  properties: {
    a: {
      description: undefined,
      type: "boolean",
    },
    b: {
      description: undefined,
      type: "number",
    },
    c: {
      description: undefined,
      type: "string",
    },
  },
  required: [
    "a",
    "b",
    "c",
  ],
  type: "object",
}
`;

snapshot[`toJSON > object 2`] = `
{
  "\$defs": {},
  additionalProperties: false,
  description: "One.",
  properties: {
    a: {
      description: undefined,
      type: "boolean",
    },
    b: {
      description: undefined,
      type: "number",
    },
    c: {
      description: undefined,
      type: "string",
    },
  },
  required: [
    "a",
    "b",
    "c",
  ],
  type: "object",
}
`;

snapshot[`toJSON > object 3`] = `
{
  "\$defs": {},
  additionalProperties: false,
  description: "One. Two.",
  properties: {
    a: {
      description: undefined,
      type: "boolean",
    },
    b: {
      description: undefined,
      type: "number",
    },
    c: {
      description: undefined,
      type: "string",
    },
  },
  required: [
    "a",
    "b",
    "c",
  ],
  type: "object",
}
`;

snapshot[`toJSON > enum 1`] = `
{
  "\$defs": {},
  description: undefined,
  enum: [
    "A",
    "B",
    "C",
  ],
  type: "string",
}
`;

snapshot[`toJSON > enum 2`] = `
{
  "\$defs": {},
  description: "One.",
  enum: [
    "A",
    "B",
    "C",
  ],
  type: "string",
}
`;

snapshot[`toJSON > enum 3`] = `
{
  "\$defs": {},
  description: "One. Two.",
  enum: [
    "A",
    "B",
    "C",
  ],
  type: "string",
}
`;

snapshot[`toJSON > union 1`] = `
{
  "\$defs": {},
  anyOf: [
    {
      description: undefined,
      type: "null",
    },
    {
      description: undefined,
      type: "string",
    },
  ],
  description: undefined,
}
`;

snapshot[`toJSON > union 2`] = `
{
  "\$defs": {},
  anyOf: [
    {
      description: undefined,
      type: "null",
    },
    {
      description: undefined,
      type: "string",
    },
  ],
  description: "One.",
}
`;

snapshot[`toJSON > union 3`] = `
{
  "\$defs": {},
  anyOf: [
    {
      description: undefined,
      type: "null",
    },
    {
      description: undefined,
      type: "string",
    },
  ],
  description: "One. Two.",
}
`;

snapshot[`toJSON > ref 1`] = `
{
  "\$defs": {},
  description: undefined,
  type: "string",
}
`;

snapshot[`toJSON > ref 2`] = `
{
  "\$defs": {},
  description: undefined,
  type: "string",
}
`;

snapshot[`toJSON > ref 3`] = `
{
  "\$defs": {},
  description: undefined,
  type: "string",
}
`;

snapshot[`toJSON > transform 1`] = `
{
  "\$defs": {},
  description: undefined,
  type: "string",
}
`;

snapshot[`toJSON > transform 2`] = `
{
  "\$defs": {},
  description: undefined,
  type: "string",
}
`;

snapshot[`toJSON > transform 3`] = `
{
  "\$defs": {},
  description: undefined,
  type: "string",
}
`;
